### ISSUE TITLE
Are structs with fixed size arrays supported?

### ISSUE THREAD

**Issue by binaryalgorithm** [m4rs-mt/ILGPU#256]
unsafe struct FSMUnit
        {
            public int tx;
            public int ty;
            public fixed int test[256];
            public fixed int test2[256];
        }
        unsafe static void MathKernel(Index2 index, ArrayView2D<FSMUnit> grid)
        {
            int tx = index.X;
            int ty = index.Y;

            grid[tx, ty].tx = tx;            
            grid[tx, ty].ty = ty;

            for (int n = 0; n < 256; n++)
            {
                grid[tx, ty].test[n] = n;
                grid[tx, ty].test2[n] = (256 - n);
            }
        }

(I am using 9.0 latest stable)

This runs except that any field after test[256] won't map correctly (even if it's just another int and not an array). Every struct fills in the tx, ty, and all 256 test values, but test2 doesn't fill in this case, except element 0 in the array gets the value "1", all other values remain at "0" (the initial value).

I have tried with annotation [StructLayout(LayoutKind.Sequential, Pack = 1)] , but this didn't change the result. It may be possible that a SubView with the correct offset can be used, but is there a reason why the struct offset mapping doesn't work after the first fixed sized array?

Also tried with:
           for (int n = 0; n < 256; n++)
            {
                grid[tx, ty].test[n] = tx;
                grid[tx, ty].test2[n] = ty;
            }

The result for say unit 64, 65 is that the test array contains all "64" values as expected, and the first element of test2 array contains "65", while the others are not set, as if it can only access the test2 array base address, but not an array index location.

It is dependent on the order of items in the struct, for example:
        unsafe struct FSMUnit
        {
            public fixed int test[256];
            public fixed int test2[256];
            public int tx;
            public int ty;
        }

^ this has the same behavior on the arrays, but tx / ty fields no longer get set at all.

**Comment by binaryalgorithm** [comment]
I setup the kernel compiler and exported the source (OpenCL 2.0) :
https://pastebin.com/LYte700x

**Comment by binaryalgorithm** [comment]
I see that it isn't using fixed array within the generated struct, and it doesn't match the expected memory layout; instead it tries to make a field for each byte generally. This might be an OpenCL specific problem with how it compiles it?

**Comment by Yey007** [comment]
I tested this first on Cuda. I can confirm that it throws a NotSupportedException when running [this code](https://gist.github.com/Yey007/b3c26fc755ad0d610c72119f3add9cb9). The exception reads:

**Comment by Yey007** [comment]
Although, I just got this exception in CL as well, so it actually running may be specific to your case?

**Comment by Yey007** [comment]
Oops, this is embarrassing... I may have been testing on 0.8.1.1. On 0.9, I can confirm the described issue on OpenCL, but on CUDA this kernel will not load at all. The exception thrown sounds a bit unrelated, but that is the nature of ILGPU exceptions at the moment:
`System.TypeLoadException: 'Access is denied: 'ILGPU.Backends.PointerViews.ViewImplementation1[ILGPUTesting.Program+FSMUnit]'.'`

In short, this is strictly a 0.9 OpenCL problem, but I have no idea where it could be. I'll definitely try looking into it and I'm sure the author/someone more experienced can offer their insight as well.

**Comment by Yey007** [comment]
I also tried with only one fixed array. Works fine on OpenCL, but on CUDA it throws the exception above. Changing the order of the arrays in the struct will also change which one gets filled. In 0.8.1.1, anything with fixed arrays appears to throw an exception, but I'm not even sure if it was implemented in that version.

**Comment by binaryalgorithm** [comment]
Looking at the compiled kernel code, it's making a field for every byte rather than an array, and the memory layout doesn't seem quite right except for the case of fixed byte[], though any of them exhibit the behavior where the first array seems to fill right and anything after that is addressed incorrectly.

I don't think the feature is properly supported. Regular structs with primitives seem to be OK though. It just happens to get some of the size information correct about the array for the OpenCL case so it will technically compile and run. A previous milestone is "Support for fixed array buffers.", but I believe this applies to local arrays inside the kernel not fixed arrays embedded in external structs. If I use explicit memory layout option on the struct, it won't be able to compile (not sure that would have helped anyway).

There might be a workaround with stored offsets and subviews, but I haven't played with that yet. Of course a workaround is to pull the arrays out of the struct and add a dimension for the unit ID, in which case I don't need any unsafe code. Looking at other projects where I was using CL kernels I wrote, fixed array mapping worked but the kernel was written with the fixed array in mind; ILGPU doesn't generate arrays like this (yet).

**Comment by Yey007** [comment]
Is the sizing actually correct? I would have expected to see 2048 char fields, not 2040. Also, what's up with the second struct with all of those char fields? Is that what gets compiled for the code you sent above?

**Comment by m4rs-mt** [comment]
@binaryalgorithm Thank you for your issue report and welcome to the community. One of the issue with fixed array buffers is the fact that the managed type information is set up in a way that we have to pad the managed structures to a particular size. ILGPU uses bytes to pad the compiler-generated types to their desired sizes (this is the source of the `byte` or `char` elements). Consider the following piece of C# code:
```c#
    public unsafe struct UnsafeStructure
    {
        public fixed int Data[1024];
    }
```
This will be internally transformed to:
```msil
.class public sequential ansi sealed beforefieldinit ConsoleApp2.UnsafeStructure
	extends [System.Runtime]System.ValueType
{
	// Nested Types
	.class nested public sequential ansi sealed beforefieldinit '<Data>e__FixedBuffer'
		extends [System.Runtime]System.ValueType
	{
		...
		.pack 0
		.size 4096

		// Fields
		.field public int32 FixedElementField

	} // end of class <Data>e__FixedBuffer

	.field public valuetype ConsoleApp2.UnsafeStructure/'<Data>e__FixedBuffer' Data
}
```
which represents another structure with a single field and a custom type size.

Since the C# compiler (and ILGPU) uses typed IO operations, this is not a compiler bug, although it looks a bit awkward 🤓. An alternative would be to create a specific pattern matcher in the internal of the compiler to emit "nicer looking" structures in the finally emitted program.

Regarding your issue: this sounds interesting since we have got a highly similar test case that works on all platforms. I am going to take a closer look 👀

**Comment by m4rs-mt** [comment]
@Yey007 the issue `TypeLoadException of ILGPUTesting.Program+FSMUnit` refers to an issue that is raised the ILGPU runtime is not able to access the type defined in this custom assembly. Maybe the `[InternalsVisibleTo...]` is missing? However, I believe that the problem is a different one 🤓

**Comment by Yey007** [comment]
@m4rs-mt  I see. So, were all those char fields in the emitted program actually data fields or were they just some sort of padding?

**Comment by Yey007** [comment]
If it was padding, where is the actual data stored in the given, compiled example? To me, it seems like type_31 should be data, not padding.Maybe I misunderstood you.

**Comment by m4rs-mt** [comment]
@Yey007 Yes and no 🤓. The field `FixedElementField` in my example above represents the first element of the fixed array buffer. Since the data structure is required to have a size of 1024 bytes, we need to add a virtual "padding" of 1020 elements of type `byte` to the structure in order to match is intended size. Since the frontend does not add padding in a "more clever" way, the byte padding will not be updated later on.

**Comment by Yey007** [comment]
@m4rs-mt I see. That makes sense now. In the compiled example, `field_0` represents the first element and all of the other char fields represent the rest of the array.

**Comment by binaryalgorithm** [comment]
The sizeof the struct reported by C# was 2048 + 8 bytes. Padding is only applied in C# when a member is not 4 byte aligned I think, but using [StructLayout(LayoutKind.Sequential, Pack = 1)] removes extra padding and makes each member adjacent to the previous one. (in this case no difference since they were all aligned).

Marshal.OffsetOf can be used to confirm the layout on the C# side:

var f1 = Marshal.OffsetOf<FSMUnit>("test");
var f2 = Marshal.OffsetOf<FSMUnit>("test2");
var f3 = Marshal.OffsetOf<FSMUnit>("tx");
var f4 = Marshal.OffsetOf<FSMUnit>("ty");

  | f1 | 0x0000000000000000 | System.IntPtr
  | f2 | 0x0000000000000400 | System.IntPtr
  | f3 | 0x0000000000000800 | System.IntPtr
  | f4 | 0x0000000000000804 | System.IntPtr

The hex offsets are correct and reflect the size of the arrays in memory. The OpenCL generated code appears to have the correct amount of bytes in the struct, so it's not clear why it can't assign values beyond the first fixed array. It is difficult for me to troubleshoot from the generated code.

**Comment by m4rs-mt** [comment]
> Padding is only applied in C# when a member is not 4 byte aligned I think

My comment was a bit misleading. From a compiler's point of view, ILGPU adds a "custom padding" to the structure in order to match its intended size of 2048 + 8 bytes 🤓. 

> The hex offsets are correct and reflect the size of the arrays in memory. 

There is an issue in the offset computation in the Cuda and CL backends. The issue is caused by multiple fixed array buffers within the same structure, which makes this an internal compiler bug.
